File: NetFramework/WinForms/CS/DotNetCore3/GaugeDemo/BaseGaugePage.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;
using C1.Win.C1Gauge;

namespace GaugeDemo
{
    public partial class BaseGaugePage : UserControl
    {
        private ToolStripMenuItem _resetMenuItem;
        private ToolStripMenuItem _categorizedMenuItem;
        private ToolStripMenuItem _alphabeticMenuItem;
        private ToolStripMenuItem _descriptionMenuItem;
        private Random _random;

        public BaseGaugePage()
        {
            InitializeComponent();
            gaugeGrid.ContextMenuStrip = CreateMenuStrip();
            _random = new Random();
        }

        //----------------------------------------------------------------------
        #region ** private

        private ContextMenuStrip CreateMenuStrip()
        {
            ContextMenuStrip result = new ContextMenuStrip();
            ToolStripSeparator separator1 = new ToolStripSeparator();
            ToolStripSeparator separator2 = new ToolStripSeparator();
            _resetMenuItem = new ToolStripMenuItem();
            _descriptionMenuItem = new ToolStripMenuItem();
            _categorizedMenuItem = new ToolStripMenuItem();
            _alphabeticMenuItem = new ToolStripMenuItem();
            result.Items.AddRange(new System.Windows.Forms.ToolStripItem[] {
                _resetMenuItem, separator1, _categorizedMenuItem, _alphabeticMenuItem,
                separator2, _descriptionMenuItem });
            _resetMenuItem.Text = "&Reset";
            _categorizedMenuItem.Text = "&Categorized";
            _alphabeticMenuItem.Text = "&Alphabetic";
            _descriptionMenuItem.Text = "&Description";
            result.Opening += new CancelEventHandler(contextMenu_Opening);
            _resetMenuItem.Click += new System.EventHandler(resetMenuItem_Click);
            _categorizedMenuItem.Click += new System.EventHandler(categorizedMenuItem_Click);
            _alphabeticMenuItem.Click += new System.EventHandler(alphabeticMenuItem_Click);
            _descriptionMenuItem.Click += new System.EventHandler(descriptionMenuItem_Click);
            return result;
        }

        private void contextMenu_Opening(object sender, CancelEventArgs e)
        {
            bool resetEnabled = false;
            GridItem item = gaugeGrid.SelectedGridItem;
            if (item != null && item.GridItemType == GridItemType.Property)
            {
                GridItem parent = item.Parent;
                PropertyDescriptor pd = item.PropertyDescriptor;
                if (parent.GridItemType == GridItemType.Root || parent.GridItemType == GridItemType.Category)
                    resetEnabled = pd.CanResetValue(gaugeGrid.SelectedObject);
                else if (parent.GridItemType == GridItemType.Property)
                    resetEnabled = pd.CanResetValue(parent.Value);
            }
            _resetMenuItem.Enabled = resetEnabled;
            _descriptionMenuItem.Checked = gaugeGrid.HelpVisible;
            _categorizedMenuItem.Checked = gaugeGrid.PropertySort == PropertySort.CategorizedAlphabetical;
            _alphabeticMenuItem.Checked = gaugeGrid.PropertySort == PropertySort.Alphabetical;
        }

        private void resetMenuItem_Click(object sender, System.EventArgs e)
        {
            GridItem item = gaugeGrid.SelectedGridItem;
            if (item != null && item.GridItemType == GridItemType.Property)
            {
                GridItem parent = item.Parent;
                PropertyDescriptor pd = item.PropertyDescriptor;
                object obj = null;
                object rootObj = gaugeGrid.SelectedObject;
                PropertyDescriptor rootPd = null;
                if (parent.GridItemType == GridItemType.Root || parent.GridItemType == GridItemType.Category)
                    obj = rootObj;
                else if (parent.GridItemType == GridItemType.Property)
                {
                    rootPd = parent.PropertyDescriptor;
                    obj = parent.Value;
                }
                if (obj != null)
                {
                    pd.ResetValue(obj);
                }
                gaugeGrid.Refresh();
            }
        }

        private void descriptionMenuItem_Click(object sender, System.EventArgs e)
        {
            gaugeGrid.HelpVisible = !gaugeGrid.HelpVisible;
        }

        private void categorizedMenuItem_Click(object sender, System.EventArgs e)
        {
            gaugeGrid.PropertySort = PropertySort.CategorizedAlphabetical;
        }

        private void alphabeticMenuItem_Click(object sender, System.EventArgs e)
        {
            gaugeGrid.PropertySort = PropertySort.Alphabetical;
        }

        private void gaugeGrid_SelectedObjectsChanged(object sender, EventArgs e)
        {
            string s = string.Empty;
            object obj = gaugeGrid.SelectedObject;
            if (obj != null)
            {
                string name = obj.GetType().ToString();
                int n = name.LastIndexOf('.');
                if (n != -1)
                {
                    s = name.Substring(n + 1) + " properties:";
                }
            }
            objectLabel.Text = s;
        }

        private void timer1_Tick(object sender, EventArgs e)
        {
            if (!IsDisposed && Visible)
            {
                OnTimer();
            }
        }

        #endregion

        internal void UpdateValue(C1GaugeBase gauge, double step)
        {
            double v = gauge.Value;
            double min, max;
            if (gauge.IsLogarithmic)
            {
                double logLogBase = Math.Log(gauge.LogarithmicBase);
                min = Math.Max(Math.Log(gauge.Minimum) / logLogBase, Math.Log(v) / logLogBase - step);
                max = Math.Min(Math.Log(gauge.Maximum) / logLogBase, Math.Log(v) / logLogBase + step);
                gauge.Value = Math.Exp((_random.NextDouble() * (max - min) + min) * logLogBase);
            }
            else
            {
                min = Math.Max(gauge.Minimum, v - step);
                max = Math.Min(gauge.Maximum, v + step);
                gauge.Value = _random.NextDouble() * (max - min) + min;
            }
        }

        internal void UpdateValue(C1GaugeBase gauge, int pointerIndex, double step)
        {
            double v = gauge.MorePointers[pointerIndex].Value;
            double min = Math.Max(gauge.Minimum, v - step);
            double max = Math.Min(gauge.Maximum, v + step);
            gauge.MorePointers[pointerIndex].Value = _random.NextDouble() * (max - min) + min;
        }

        internal void UpdateValue(C1GaugeBase gauge, string pointerName, double step, double min, double max)
        {
            C1GaugePointer pointer = gauge.MorePointers[pointerName];
            min = Math.Max(min, pointer.Value - step);
            max = Math.Min(max, pointer.Value + step);
            pointer.Value = _random.NextDouble() * (max - min) + min;
        }

        protected virtual void OnTimer()
        {
        }
    }
}

File: NetFramework/Gauge/VB/GaugeDemo/BaseGaugePage.vb
Imports System.ComponentModel
Imports C1.Win.C1Gauge

Public Class BaseGaugePage

    Private _resetMenuItem As ToolStripMenuItem
    Private _categorizedMenuItem As ToolStripMenuItem
    Private _alphabeticMenuItem As ToolStripMenuItem
    Private _descriptionMenuItem As ToolStripMenuItem
    Private _random As Random

    Public Sub New()
        MyBase.New()

        InitializeComponent()

        gaugeGrid.ContextMenuStrip = CreateMenuStrip()
        _random = New Random()

    End Sub

#Region "** private"

    Private Function CreateMenuStrip() As ContextMenuStrip
        Dim result As ContextMenuStrip = New ContextMenuStrip()
        Dim separator1 As ToolStripSeparator = New ToolStripSeparator()
        Dim separator2 As ToolStripSeparator = New ToolStripSeparator()
        _resetMenuItem = New ToolStripMenuItem()
        _descriptionMenuItem = New ToolStripMenuItem()
        _categorizedMenuItem = New ToolStripMenuItem()
        _alphabeticMenuItem = New ToolStripMenuItem()
        result.Items.AddRange(New System.Windows.Forms.ToolStripItem() { _
            _resetMenuItem, separator1, _categorizedMenuItem, _alphabeticMenuItem, _
            separator2, _descriptionMenuItem})
        _resetMenuItem.Text = "&Reset"
        _categorizedMenuItem.Text = "&Categorized"
        _alphabeticMenuItem.Text = "&Alphabetic"
        _descriptionMenuItem.Text = "&Description"
        AddHandler result.Opening, AddressOf contextMenu_Opening
        AddHandler _resetMenuItem.Click, AddressOf resetMenuItem_Click
        AddHandler _categorizedMenuItem.Click, AddressOf categorizedMenuItem_Click
        AddHandler _alphabeticMenuItem.Click, AddressOf alphabeticMenuItem_Click
        AddHandler _descriptionMenuItem.Click, AddressOf descriptionMenuItem_Click
        Return result
    End Function

    Private Sub contextMenu_Opening(ByVal sender As Object, ByVal e As CancelEventArgs)
        Dim resetEnabled As Boolean = False
        Dim item As GridItem = gaugeGrid.SelectedGridItem
        If item IsNot Nothing AndAlso item.GridItemType = GridItemType.Property Then
            Dim parent As GridItem = item.Parent
            Dim pd As PropertyDescriptor = item.PropertyDescriptor
            If parent.GridItemType = GridItemType.Root Or parent.GridItemType = GridItemType.Category Then
                resetEnabled = pd.CanResetValue(gaugeGrid.SelectedObject)
            ElseIf parent.GridItemType = GridItemType.Property Then
                resetEnabled = pd.CanResetValue(parent.Value)
            End If
        End If
        _resetMenuItem.Enabled = resetEnabled
        _descriptionMenuItem.Checked = gaugeGrid.HelpVisible
        If gaugeGrid.PropertySort = PropertySort.CategorizedAlphabetical Then
            _categorizedMenuItem.Checked = True
        Else
            _categorizedMenuItem.Checked = False
        End If
        If gaugeGrid.PropertySort = PropertySort.Alphabetical Then
            _alphabeticMenuItem.Checked = True
        Else
            _alphabeticMenuItem.Checked = False
        End If
    End Sub

    Private Sub resetMenuItem_Click(ByVal sender As Object, ByVal e As System.EventArgs)
        Dim item As GridItem = gaugeGrid.SelectedGridItem
        If (item IsNot Nothing AndAlso item.GridItemType = GridItemType.Property) Then
            Dim parent As GridItem = item.Parent
            Dim pd As PropertyDescriptor = item.PropertyDescriptor
            Dim obj As Object = Nothing
            Dim rootObj As Object = gaugeGrid.SelectedObject
            Dim rootPd As PropertyDescriptor = Nothing
            If Parent.GridItemType = GridItemType.Root Or Parent.GridItemType = GridItemType.Category Then
                obj = rootObj
            ElseIf Parent.GridItemType = GridItemType.Property Then
                rootPd = Parent.PropertyDescriptor
                obj = Parent.Value
            End If
            If obj IsNot Nothing Then
                pd.ResetValue(obj)
            End If
            gaugeGrid.Refresh()
        End If
    End Sub

    Private Sub descriptionMenuItem_Click(ByVal sender As Object, ByVal e As System.EventArgs)
        gaugeGrid.HelpVisible = Not gaugeGrid.HelpVisible
    End Sub

    Private Sub categorizedMenuItem_Click(ByVal sender As Object, ByVal e As System.EventArgs)
        gaugeGrid.PropertySort = PropertySort.CategorizedAlphabetical
    End Sub

    Private Sub alphabeticMenuItem_Click(ByVal sender As Object, ByVal e As System.EventArgs)
        gaugeGrid.PropertySort = PropertySort.Alphabetical
    End Sub

    Private Sub gaugeGrid_SelectedObjectsChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles gaugeGrid.SelectedObjectsChanged
        Dim s As String = String.Empty
        Dim obj As Object = gaugeGrid.SelectedObject
        If obj IsNot Nothing Then
            Dim name As String = obj.GetType().ToString()
            Dim n As Int32 = name.LastIndexOf(".")
            If n <> -1 Then
                s = name.Substring(n + 1) + " properties:"
            End If
        End If
        objectLabel.Text = s
    End Sub

    Private Sub timer1_Tick(ByVal sender As Object, ByVal e As System.EventArgs) Handles timer1.Tick
        If Not IsDisposed AndAlso Visible Then
            OnTimer()
        End If
    End Sub

#End Region

    Friend Sub UpdateValue(ByVal gauge As C1GaugeBase, ByVal maxStep As Double)
        Dim v As Double = gauge.Value
        Dim min, max As Double
        If (gauge.IsLogarithmic) Then
            Dim logLogBase As Double = Math.Log(gauge.LogarithmicBase)
            min = Math.Max(Math.Log(gauge.Minimum) / logLogBase, Math.Log(v) / logLogBase - maxStep)
            max = Math.Min(Math.Log(gauge.Maximum) / logLogBase, Math.Log(v) / logLogBase + maxStep)
            gauge.Value = Math.Exp((_random.NextDouble() * (max - min) + min) * logLogBase)
        Else
            min = Math.Max(gauge.Minimum, v - maxStep)
            max = Math.Min(gauge.Maximum, v + maxStep)
            gauge.Value = _random.NextDouble() * (max - min) + min
        End If
    End Sub

    Friend Sub UpdateValue(ByVal gauge As C1GaugeBase, ByVal pointerIndex As Int32, ByVal maxStep As Double)
        Dim v As Double = gauge.MorePointers(pointerIndex).Value
        Dim min As Double = Math.Max(gauge.Minimum, v - maxStep)
        Dim max As Double = Math.Min(gauge.Maximum, v + maxStep)
        gauge.MorePointers(pointerIndex).Value = _random.NextDouble() * (max - min) + min
    End Sub

    Friend Sub UpdateValue(ByVal gauge As C1GaugeBase, ByVal pointerName As String, ByVal maxStep As Double, ByVal min As Double, ByVal max As Double)
        Dim pointer As C1GaugePointer = gauge.MorePointers(pointerName)
        min = Math.Max(min, pointer.Value - maxStep)
        max = Math.Min(max, pointer.Value + maxStep)
        pointer.Value = _random.NextDouble() * (max - min) + min
    End Sub

    Protected Overridable Sub OnTimer()
    End Sub

End Class
